<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.56">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>galaxy – Introduction to NGS data analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-EXW13D5861"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-EXW13D5861', { 'anonymize_ip': true});
</script>
<script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "TrainingMaterial",
    "@id": "https://hds-sandbox.github.io/NGS_summer_course_Aarhus/",
    "dct:conformsTo": {
      "http://purl.org/dc/terms/conformsTo": {
        "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
        "@type": "CreativeWork"
      }
    },
    "name": "Introduction to NGS data analysis",
    "description": "A one-week summer school which provides an overview of current NGS technologies and directions in which these are currently progressing. Methods for assembling genomes from NGS data using either a reference genome or de novo assembly will be presented together with some emphasis on assessing NGS data quality (Quality control, accuracy of base calling). The course includes notebooks for the analysis of short- and long-read sequencing data, bulkRNA data and single-cell RNA sequencing data.",
    "keywords": [
      "single cell RNA",
      "NGS data",
      "genomics",
      "transcriptomics",
      "bulk RNA",
      "bioinformatics"
    ],
    "about": [
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/topic_0091",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_0091",
        "name": "bioinformatics",
        "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Ftopic_0091"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/data_3670",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "data_3670",
        "name": "Online course",
        "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Fdata_3670"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/topic_0622",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_0622",
        "name": "Genomics",
        "url": "https://www.ebi.ac.uk/ols4/ontologies/edam/classes/http%253A%252F%252Fedamontology.org%252Ftopic_0622"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/topic_3308",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_3308",
        "name": "Transcriptomics",
        "url": "https://www.ebi.ac.uk/ols4/ontologies/edam/classes/http%253A%252F%252Fedamontology.org%252Ftopic_3308"
      }
    ],
    "abstract": "NGS technologies are fundamental in genomics and transcriptomics studies. In this course we set the foundations for the analysis of NGS data with various interactive exercises for genome alignment, genomics and transcriptome analysis. open source code, Slides and access information can be found at the webpage https://hds-sandbox.github.io/NGS_summer_course_Aarhus/.",
    "audience": [
      {
        "@type": "Audience",
        "audienceType": "Students"
      },
      {
        "@type": "Audience",
        "audienceType": "PhD students"
      },
      {
        "@type": "Audience",
        "audienceType": "Postdoctoral researchers"
      }
    ],
    "competencyRequired": [
      "Basic biology programming",
      "Basic python/R/bash programming (an advantage but not mandatory)",
      "Machine learning basics for bioinformatics (PCA, statistical testing)"
    ],
    "inLanguage": ["en-US"],
    "learningResourceType": ["tutorials"],
    "license": "https://spdx.org/licenses/CC-BY-4.0.html",
    "url": "https://hds-sandbox.github.io/NGS_summer_course_Aarhus/",
    "citation": [
        {
            "@type": "CreativeWork",
            "@id": "https://doi.org/10.5281/zenodo.8074603"
        } ],
     "funding":[
     	  {
            "@type": "Grant",
            "@id": "NNF grant number NNF20OC0063268"        	
        }
     ]
  }
</script>


<meta property="og:title" content="Introduction to NGS data analysis">
<meta property="og:description" content="">
<meta property="og:site_name" content="Introduction to NGS data analysis">
<meta property="og:locale" content="en_US">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/NGSlogo.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Introduction to NGS data analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../slides.html"> 
<span class="menu-text">Slides</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-access" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Access</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-access">    
        <li>
    <a class="dropdown-item" href="../../access/ngssummer.html">
 <span class="dropdown-text">Summer School</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../access/UCloud.html">
 <span class="dropdown-text">UCloud</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../access/genomedk.html">
 <span class="dropdown-text">GenomeDK</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-tutorials" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Tutorials</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-tutorials">    
        <li>
    <a class="dropdown-item" href="../../galaxy/galaxy.html">
 <span class="dropdown-text">Alignment with Galaxy</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../nb/01_raw_data_alignment.html">
 <span class="dropdown-text">Alignment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../nb/02_VCF_variant_calling_analysis.html">
 <span class="dropdown-text">Variant analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../nb/03_bulk_RNA_analysis.html">
 <span class="dropdown-text">bulkRNA analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../nb/04_scRNAseq_analysis.html">
 <span class="dropdown-text">scRNA analysis</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/hds-sandbox"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#quality-control-and-alignment-of-raw-reads" id="toc-quality-control-and-alignment-of-raw-reads" class="nav-link active" data-scroll-target="#quality-control-and-alignment-of-raw-reads">Quality Control and Alignment of raw reads</a>
  <ul class="collapse">
  <li><a href="#biological-introduction" id="toc-biological-introduction" class="nav-link" data-scroll-target="#biological-introduction">Biological introduction</a></li>
  <li><a href="#exercise-guide" id="toc-exercise-guide" class="nav-link" data-scroll-target="#exercise-guide">Exercise guide</a>
  <ul class="collapse">
  <li><a href="#prepare-the-tools-and-data" id="toc-prepare-the-tools-and-data" class="nav-link" data-scroll-target="#prepare-the-tools-and-data">Prepare the tools and data</a></li>
  <li><a href="#galaxy-workflow" id="toc-galaxy-workflow" class="nav-link" data-scroll-target="#galaxy-workflow">Galaxy Workflow</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">




<section id="quality-control-and-alignment-of-raw-reads" class="level1">
<h1>Quality Control and Alignment of raw reads</h1>
<p>This first exercise will be executed on <code>Galaxy</code>, an interactive platform to run bioinformatics workflows. We will replicate this lesson with computer code later in this course. <code>Galaxy</code> has the possibility of working with a free account.</p>
<section id="biological-introduction" class="level2">
<h2 class="anchored" data-anchor-id="biological-introduction">Biological introduction</h2>
<p>White clover (<em>Trifolium repens</em>) is an <strong>allotetraploid</strong>. This means that it contains genomes originating from two different species within the same nucleus. Normally, white clover is an outbreeding species, but a self-compatible line was used for <a href="https://academic.oup.com/plcell/article/31/7/1466/5985684">sequencing the white clover genome</a>. This line is designated <code>S10</code> in your data, indicating that this is the 10th self-fertilized generation. In addition, you have data from a wild clover accession (ecotype) called Tienshan (<code>Ti</code>), which is collected from Chinese mountains and is adapted to alpine conditions.</p>
<p><img src="./img/clover.png" class="img-fluid"> <em>Figure: Characterisation of the white clover population. T.Repens is a hybrid of T.Occidentale and T.Pallescens</em></p>
</section>
<section id="exercise-guide" class="level2">
<h2 class="anchored" data-anchor-id="exercise-guide">Exercise guide</h2>
<section id="prepare-the-tools-and-data" class="level3">
<h3 class="anchored" data-anchor-id="prepare-the-tools-and-data">Prepare the tools and data</h3>
<ol type="1">
<li>Install IGV on your computer from <a href="https://software.broadinstitute.org/software/igv/download">here</a>. This is a genome browser you will use to look at some files.</li>
<li>Create an account at <a href="https://usegalaxy.org">usegalaxy.org</a> and log into galaxy.</li>
<li>Find the course data by going to <a href="https://usegalaxy.org/u/samuele.soraggi/h/copy-of-ngs-2023-course">this web address</a> and by clicking on <code>Import this history</code> (top left corner of the page).</li>
</ol>
<p><strong>Note:</strong> If usegalaxy.org has availability problems, you can use the other server <a href="https://usegalaxy.eu/">https://usegalaxy.eu/</a> and get the data at <a href="https://usegalaxy.eu/u/samuele.soraggi/h/ngs2023">this link</a>. You might need to create an account there as well.</p>
<section id="data" class="level4">
<h4 class="anchored" data-anchor-id="data">Data</h4>
<p>You will be working with two types of sequencing data. The first is <em>PacBio Hifi reads</em>, which are long and accurate. You can find them under Hifi_reads_white_clover.fastq. The second type is <em>Illumina RNA-seq reads</em>, which are short and accurate and should be aligned using <a href="https://youtu.be/XbBS9o-4100">a spliced aligner, such as STAR</a>.</p>
<p>There are 24 of these files, 12 for each of the two genotypes mentioned before. The files are named <code>[genotype]_[treatment]_[replicate].fastq</code>. Treatment 1 is before and treatment 2 is after exposure to frost, respectively.</p>
<p>In addition to the sequencing data, there are also three reference files: one for homologous contig 1 (referencing T. occidentale-derived subgenome), one for contig 2 (T. pallescens-derived subgenome) and one for both Contigs 1 and 2. The reference files are in <code>fasta</code> format.</p>
<p>The file <code>white_clover_genes.gtf</code> contains the gene annotations for the two contigs.</p>
</section>
<section id="workflow-illustration" class="level4">
<h4 class="anchored" data-anchor-id="workflow-illustration">Workflow illustration</h4>
<p>Through <code>Galaxy</code>, we build a workflow applying tools to the data. We will look at the quality of the raw reads for both PacBio HiFi and Illumina RNA-seq reads. Afterwards, we align to references, using two different tools for the two types of data. Finally, we will look at the alignments on a genome browser. We will work then on a computing cluster through <code>uCloud</code> to analyze the aligned data in some of the upcoming lessons of the course.</p>
<p><img src="./img/scheme_exercise.png" class="img-fluid"></p>
</section>
</section>
<section id="galaxy-workflow" class="level3">
<h3 class="anchored" data-anchor-id="galaxy-workflow">Galaxy Workflow</h3>
<p>When you import the files, what you actually import is a <strong>History</strong> - a sequence of files and softwares applied on the data. You can see the history on the right side of your <em>usegalaxy.org</em> webpage with green panels. Here, we only have the starting data, and you will build the rest of your history through various tools.</p>
<p><img src="./img/history.png" class="img-fluid"></p>
<p>On the left side of the screen, you have a menu with various available tools organized by category. All those softwares are also available on a classical computing command line (we will try those as well).</p>
<p><img src="./img/tools.png" class="img-fluid"></p>
<section id="quality-control" class="level4">
<h4 class="anchored" data-anchor-id="quality-control">Quality control</h4>
<p><strong>1</strong> Run FastQC on the PacBio Hifi reads and on two of the Illumina RNA-seq libraries. FastQC does quality control of the raw sequence data, providing an overview of the data which can help identify if there are any problems that should be addressed before further analysis.</p>
<p>In the tool menu, click on <code>FASTQ quality control --&gt; FASTQC read quality reports</code>. You will see a window with <em>tool parameters</em>: for the first option (raw read data from history), choose multiple files and select <code>Hifi_reads_white_clover.fastq</code> plus other <code>fastq</code> files you want to see the quality of (example in figure below). Then click on the button <code>Run Tool</code>. <img src="./img/fastqc.png" class="img-fluid"></p>
<p>You will notice that some new elements are added to your history. Part of them are <code>FastQC</code> producing a text file, while others are <code>FastQC</code> producing a webpage report. The reports are ready when coloured in green: click on the <em>eye symbol</em> of a history item to read a report.</p>
<p><strong>2</strong> FastQC provides a report for each sample. To have a better comparison between the <em>Hifi</em> and <em>Illumina</em> data, we would combine the three <code>FastQC</code> reports into one using <code>MultiQC</code>.</p>
<p>Choose the MultiQC tool from <code>FASTQ quality control --&gt; MultiQC aggregate results from ...</code>. In the options, select <code>FastQC</code> as the used tool for the logsselect FastQC as the tool used to generate the output, and then select the items of <code>FastQC</code> of your history producing <code>RawData</code> (Figure below). In this way, you build a pipeline from the previous reports to the new tool you are using. Now click on <code>Run Tool</code>.</p>
<p><img src="./img/multiqc.png" class="img-fluid"></p>
<p>The tool will be now running in your history. When it is done, click on the <em>eye symbol</em> to see the report.</p>
<p>select the three “RawData” outputs generated by FastQC. Visualize the Webpage generated by MultiQC. Hint: You can find a “Help” button that offers additional information about the plots for each panel.</p>
<summary>
Questions:
<p>
<i>Focus on the following panels: “Per base sequence quality”, “Per sequence quality scores”…. (“Per base sequence content” always gives a FAIL for RNA-seq data). What do you notice with respect to the sequence quality scores? And are there any other quality issues worth noting? </i>
</p>
</summary>
</section>
<section id="hifi-data-alignment" class="level4">
<h4 class="anchored" data-anchor-id="hifi-data-alignment">Hifi Data Alignment</h4>
<p><strong>3</strong> Map the PacBio Hifi reads (<code>Hifi_reads_white_clover.fastq</code>) to the white clover reference sequence (Contigs 1 and 2) using <code>minimap2</code> (Map with minimap2). Find <code>Genomics Analysis --&gt; Mapping --&gt; Map with minimap2</code>. In the options, do not leave <code>Use a built-in genome index</code>, but select the option for having a genome from history. Choose then <code>DNA_Contig1_2.fasta</code> as the reference sequence.</p>
<p>Under the profile with preset options, choose <code>PacBio/Oxford Nanopore read to reference mapping (map-pb)</code>. Then click on <code>Run Tool</code>.</p>
<p><img src="./img/map-pb.png" class="img-fluid"></p>
<p><strong>4</strong> Run the same alignment, but choose as preset options <code>Long assembly to reference mapping. Divergence is far below 20% (asm20)</code>.</p>
<p>Rename then the two alignments using the edit function (<em>pen symbol</em> in the history). Use for example names <code>Contig1_2_mappb</code> and <code>Contig1_2_asm20</code>, to distinguish alignment options and reference genome.</p>
<p><strong>5</strong> The aligned genomes are not sorted by coordinates. Sort the alignments using <code>Samtools sort</code> (Find the tool under <code>Genomic file manipulation --&gt; SAM/BAM --&gt; Samtools sort ...</code>). In the options, choose the two aligned files with multiple selection. Then click on <code>Run Tool</code>.</p>
<p><strong>6</strong> Download the two alignments to your computer. To do so, click on the <em>disk symbol</em> of each file in your history, and for each download both the Dataset (alignments in <code>bam</code> format) and their index files (in <code>bai</code> format). Download as well the reference genome in <code>fasta</code> format (<code>DNA_Contig1_2.fasta</code> from the history).</p>
<p><strong>7</strong> Open IGV on your computer. Load the reference first: go on <code>Genome --&gt; Load genome from file</code> and select the <code>fasta</code> file you downloaded. Then load the two alignments: go on <code>File --&gt; Load from file</code> and select the <code>bam</code> and <code>bai</code> files you downloaded, together. You can now visualize the alignments.</p>
<p><img src="./img/IGV.png" class="img-fluid"></p>
<summary>
Questions:
<p>
<i> Look at the alignments in IGV. What do you notice about the alignments? What is the difference between the two alignments? Do you think one of them is better than the other? Choose on of the two alignments for the next steps. </i>
</p>
</summary>
<p><strong>8</strong> Repeat the alignment with Minimap2 (using the chosen alignment option from the question above) and the sorting, but using the reference genomes for Contig 1 and for Contig 2 searaately. <strong>Note:</strong> you can run all at once by choosing multiple reference genomes in the options!</p>
<summary>
Questions:
<p>
</p><p><i> Download the two references for Contig 1 and 2, and the two sorted alignments. Load the references from the menu <code>Genomes</code> in IGV, and then open the two alignments using the menu <code>File</code> in IGV.</i></p><i>
</i><ul><i>
<li><p>Why do you see fluctuations in coverage and large regions without any apparent subgenome SNPs?</p></li>
</i><li><i></i><p><i>What are the major differences between the stats for the reads mapped to Contigs1&amp;2 versus contig1 and contig2? What is your interpretation of the differences? </i></p></li>
</ul>
<p></p>
</summary>
</section>
<section id="rna-data-alignment" class="level4">
<h4 class="anchored" data-anchor-id="rna-data-alignment">RNA Data Alignment</h4>
<p><strong>9</strong> First, group the 24 RNA-seq libraries into two dataset lists, one list of pairs for S10 libraries and another for Tienshan libraries. so we can work with multiple samples simultaneously. You can do this by selecting the libraries for each genotype and choosing <code>Build Lists of Dataset Pairs</code>.</p>
<p><img src="./img/buildlist.png" class="img-fluid"></p>
<p>The grouping suggested by Galaxy is wrong, because the paired reads are paired according to <code>_1</code> and <code>_2</code> in their names. Change those two with <code>R1</code> and <code>R2</code>, click on <code>Unpair all</code>, and then click on the suggested corrected pairs with the buttons <code>Pair these datasets</code>.</p>
<p><img src="./img/groupoptions.png" class="img-fluid"></p>
<p>Your sequences will be substituted by two elements in your history. Here we chose for example to name them <code>S10</code> and <code>TI</code>.</p>
<p><img src="./img/builtlist.png" class="img-fluid"></p>
<p><strong>10</strong> Do alignment of the RNA-seq lists of raw files to the reference <code>DNA_Contig1_2.fasta</code> using <code>STAR</code>. Go to <code>Genomics analysis --&gt; RNA-seq --&gt; RNA STAR Gapped-read mapper for RNA-seq data</code>. In the options use:</p>
<ul>
<li>as data, the parameter <code>Paired-end (as collection)</code>, and then choose one of the two collections (you cannot run them all at once)</li>
<li>as reference, <code>DNA_Contig1_2.fasta</code>, with Length of SA pre-indexing string equal to <code>9</code></li>
<li>as index with gene-model, use <code>white_clover_annotations.gtf</code></li>
<li>as output, <code>Per gene read counts (GeneCounts)</code>.</li>
</ul>
<p><strong>11</strong> Use <code>MultiQC</code> to see the quality of the output. The alignment of <code>STAR</code> produces log files which can be used for quality reports. Go on <code>Genomic File Manipulation --&gt; MultiQC</code>. In the options select the tool <code>STAR</code>. Then <code>Insert STAR output</code>, as type of output the <code>Log</code>, and choose the two logs listing collections of <code>STAR</code> alignments. Then click on <code>Run Tool</code>.</p>
<p><img src="./img/starMultiQC.png" class="img-fluid"></p>
<p>View the report to see the alignment statistics.</p>
<p><strong>Final note:</strong> `Galaxy can also be used to create an automatic workflow that will map the data. This workflow can be useful when running multiple samples. You can generate a workflow from the analysis already completed in a history, by going to Settings → Extract workflow. You can also create a workflow from scratch using the Workflow editor.</p>


</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/hds-sandbox\.github\.io\/NGS_summer_course_Aarhus\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>